trieste/acquisition/function/function.py

```diff
@@ -1034,6 +1035,608 @@ def __call__(self, x: TensorType) -> TensorType:
+
+        R_whole = Sigma_uv - Sigma_iu * Sigma_iv / Sigma_ii
+
+        def create_blocks(q: int) -> TensorType:
+
+            block1 = tf.concat(
+                [
+                    R_whole[:, q, :q, :q],
+                    R_whole[:, q, q + 1 :, :q],
+                ],
+                axis=1,
+            )
+
+            block2 = tf.concat(
+                [
+                    R_whole[:, q, :q, q + 1 :],
+                    R_whole[:, q, q + 1 :, q + 1 :],
+                ],
+                axis=1,
+            )
+
+            R_block = tf.concat([block1, block2], axis=2)
+
+            return R_block
+
+        R = tf.map_fn(
+            create_blocks,
+            tf.range(Q),
+            fn_output_signature=R_whole.dtype,
+        )
+        R = tf.transpose(R, perm=[1, 0, 2, 3])
+
+        return R
+
+    def compute_Phi(
+        self,
+        c: tf.Tensor,
+        R: tf.Tensor,
+        mvn_cdf: Callable[[TensorType, TensorType, TensorType, float], TensorType],
+    ) -> TensorType:
+        """Helper function for the batch expected improvement, which computes
+        the tensor Phi, which is the tensor of multivariate Gaussian CDFs, in
+        the inner sum of the equation (3) in Chevallier and Ginsbourger
+        :cite:`chevalier2013fast`.
+
+        :param c: Tensor of shape (BQ, Q, Q-1).
+        :param R: Tensor of shape (BQ, Q, Q-1, Q-1).
+        :param mvn_cdf: Multivariate Gaussian CDF, made using MultivariateNormalCDF.
+        :returns Phi: Tensor of multivariate Gaussian CDFs.
+        """
+
+        # Check shapes of covariance tensor
+        tf.debugging.assert_shapes(
+            [
+                (c, ("BQ", "Q", "Q_")),
+                (R, ("BQ", "Q", "Q_", "Q_")),
+            ]
+        )
+
+        # Unpack tensor shape and data type
+        BQ, Q, _, Q_ = R.shape
+        dtype = R.dtype
+
+        try:
+            assert BQ % Q == 0
+
+        except AssertionError:
+            raise ValueError(
+                f"Expected size of dimension 0 of R tensor to be "
+                f"divisible by size of dimension 1, instead found "
+                f"{R.shape[0]} and {R.shape[1]}."
+            )
+
+        # Compute parallelisation dimension from batch size
+        B = BQ // Q
+
+        c_reshaped = tf.reshape(c, (BQ * Q, Q - 1))
+        R_reshaped = tf.reshape(R, (BQ * Q, Q - 1, Q - 1))
+
+        # Compute mean, covariance and x for Phi mvn normal cdf
+        Phi_cdf_x = c_reshaped  # (B*Q, Q-1)
+        Phi_cdf_mean = tf.zeros(shape=(BQ * Q, Q - 1), dtype=dtype)  # (B*Q*Q, Q)
+        Phi_cdf_cov = R_reshaped  # (B*Q*Q, Q-1, Q-1)
+
+        # Compute multivariate cdfs
+        mvn_cdfs = mvn_cdf(  # type: ignore
+            x=Phi_cdf_x,
+            mean=Phi_cdf_mean,
+            cov=Phi_cdf_cov,
+        )
+        mvn_cdfs = tf.reshape(mvn_cdfs, (B, Q, Q))  # (B, Q, Q)
+
+        return mvn_cdfs
+
+    def compute_batch_expected_improvement(
```

Comment:
can you have leading dimensions with this? i.e. mean is `[..., B, Q]`

would probably require changing a number of things but I don't see any blocker?

Reply:
One simple way to achieve this is for the user to do a reshape before calling the method:

``` 
# Assume:
#     mean.shape == [..., B, Q]
#     covariance == [..., B, Q, Q]
#     threshold == [..., B]
# Then reshape:

mean = tf.reshape(mean, shape=(-1, Q))
covariance = tf.reshape(covariance, shape=(-1, Q, Q))

# and call the function as you would normally
```

alternatively, we can do this in the function itself. Which do you think is best?